src/BinSkim.Rules/MachORules/BA5001.EnablePositionIndependentExecutable.cs

```diff
@@ -0,0 +1,107 @@
+        ///  This can be enabled by passing '-f pie' to clang/gcc."
+        /// </summary>
+        public override MultiformatMessageString FullDescription => new MultiformatMessageString { Text = RuleResources.BA5001_EnablePositionIndependentExecutable_Description };
+
+        protected override IEnumerable<string> MessageResourceNames => new string[] {
+                    nameof(RuleResources.BA5001_Pass),
+                    nameof(RuleResources.BA5001_Error),
+                    nameof(RuleResources.NotApplicable_InvalidMetadata)
+                };
+
+        public override AnalysisApplicability CanAnalyze(MachOBinary target, Sarif.PropertiesDictionary policy, out string reasonForNotAnalyzing)
+        {
+            reasonForNotAnalyzing = null;
+            if (target is MachOBinary mainMachO)
+            {
+                foreach (SingleMachOBinary singleMachO in mainMachO.MachOs)
+                {
+                    if (IsApplicableMachO(singleMachO))
+                    {
+                        // if any macho is applicable
+                        return AnalysisApplicability.ApplicableToSpecifiedTarget;
+                    }
+                }
+
+                reasonForNotAnalyzing = MetadataConditions.MachOIsNotExecutableDynamicLibraryOrObject;
+            }
+
+            return AnalysisApplicability.NotApplicableToSpecifiedTarget;
+        }
+
+        public override void Analyze(BinaryAnalyzerContext context)
+        {
+            IDwarfBinary binary = context.DwarfBinary();
+            if (binary is MachOBinary mainBinary)
+            {
+                foreach (SingleMachOBinary subBinary in mainBinary.MachOs)
+                {
+                    if (IsApplicableMachO(subBinary) && !HasPieFlag(subBinary))
+                    {
+                        context.Logger.Log(this,
+                            RuleUtilities.BuildResult(FailureLevel.Error, context, null,
+                                nameof(RuleResources.BA5001_Error),
+                                context.TargetUri.GetFileName()));
+                        return;
+                    }
+                }
+
+                // PIE enabled on executable '{0}'.
+                context.Logger.Log(this,
+                    RuleUtilities.BuildResult(ResultKind.Pass, context, null,
+                        nameof(RuleResources.BA5001_Pass),
+                        context.TargetUri.GetFileName()));
+            }
+        }
+
+        private static bool IsApplicableMachO(SingleMachOBinary macho)
+        {
+            return (macho.MachO.FileType == ELFSharp.MachO.FileType.Executable ||
+                    macho.MachO.FileType == ELFSharp.MachO.FileType.DynamicLibrary);
+        }
+
+        private static bool HasPieFlag(SingleMachOBinary binary)
+        {
+            if (binary.MachO.FileType == ELFSharp.MachO.FileType.DynamicLibrary)
+            {
+                // for shared libaries the compiler adds -fPIC by default
```

Comment:
>fPIC [](http://example.com/codeflow?start=58&length=4)

what if u have a sharedlibrary and manually remove? what would happen? is it possible?

Reply:
yes PIC can be disabled by option `-mdynamic-no-pic`
https://developer.apple.com/library/archive/documentation/DeveloperTools/Conceptual/MachOTopics/1-Articles/dynamic_code.html 

Also there is a check in https://liquid.microsoft.com/Web/Object/Read/MS.Security/Requirements/Microsoft.Security.SystemsADM.10203#Zguide says
>Do not enable any of the following flags:
>
>Allow Executable Stack (--allow_stack_execute)
>Remove PIC References (--mdynamic-no-pic)

Should we keep them in 2 rules or check both condition in current rule?

Reply:
The point here is: we should always check even if this is a dynamiclibrary or not.
Because, as you mentioned, anyone could disable it and if thats disabled, u would need to flag it, right?

Reply:
I see, the mach-O libraries don't have a flag similar to executables PIE flag in file header. Need to investigate more about how to tell if library is built with -fPIC option.

Reply:
I double checked the Apple's article about PIC https://developer.apple.com/library/archive/documentation/DeveloperTools/Conceptual/MachOTopics/1-Articles/dynamic_code.html#//apple_ref/doc/uid/TP40002528

I think my previous understanding was wrong. It looks like the Position-Independent Code (PIC) technique is built-in with Mach-O
>Mach-O position-independent code design is based on the observation that the __DATA segment is always located at a constant offset from the __TEXT segment. That is, the dynamic loader, when loading any Mach-O file, never moves a file’s __TEXT segment relative to its __DATA segment. Therefore, a function can use its own current address plus a fixed offset to determine the location of the data it wishes to access. All segments of a Mach-O file, not only the __TEXT and __DATA segments, are at fixed offsets relative to the other segments.

But with option `-mdynamic-no-pic`, it still accesses indirect addressing through relocation section.
> This option both reduces the code size of application executables and improves their performance by eliminating position-independent code references,  ___while preserving indirect calls to shared libraries and indirection to undefined symbols.___

Reply:
I think we can do everything in this rule:
1. check if the user has enabled pie -> if yes, pass
2. check if the user has enabled no pie flag -> if yes, error

what do u think?